pkg/controller/cache/cache_controller.go

```diff
@@ -92,10 +111,163 @@ func (r *ReconcileCache) Reconcile(request reconcile.Request) (reconcile.Result,
 		return reconcile.Result{}, err
 	}
 
+	// Reconcile cache
 	reqLogger.Info("Identify the target cluster")
-	reqLogger.Info("If cache doesn't exist create")
-	reqLogger.Info("If cache already exists, check if it's compatible with the CR request. If not then do appropriate actions")
-	reqLogger.Info("Update CR status with connnection info")
+	// Fetch the Infinispan cluster info
+	ispnInstance := &infinispanv1.Infinispan{}
+	nsName := types.NamespacedName{Name: instance.Spec.ClusterName, Namespace: instance.Namespace}
+	err = r.client.Get(context.TODO(), nsName, ispnInstance)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			reqLogger.Error(err, "Infinispan cluster "+ispnInstance.Name+" not found")
+			return reconcile.Result{RequeueAfter: time.Second * 10}, err
+		}
+		// Error reading the object - requeue the request.
+		return reconcile.Result{}, err
+	}
+
+	// Cluster must be well formed
+	for i, c := range ispnInstance.Status.Conditions {
+		if c.Type == "wellFormed" && c.Status == "True" {
+			break
+		}
+		if i == len(ispnInstance.Status.Conditions)-1 {
+			reqLogger.Info("Infinispan cluster " + ispnInstance.Name + " not well formed")
+			return reconcile.Result{RequeueAfter: time.Second * 10}, err
+		}
+	}
+	// Authentication is required to go on from here
+	user, pass, result, err := getCredentials(r, reqLogger, instance)
+	if err != nil {
+		return result, err
+	}
+	// List the pods for this infinispan's deployment
+	podList := &corev1.PodList{}
+	labelSelector := labels.SelectorFromSet(util.GetLabelsForInfinispan(ispnInstance.Name, ""))
+	listOps := &client.ListOptions{Namespace: ispnInstance.GetClusterName(), LabelSelector: labelSelector}
+	err = r.client.List(context.TODO(), podList, listOps)
+	if err != nil || (len(podList.Items) == 0) {
+		reqLogger.Error(err, "failed to list pods", "Infinispan.Namespace")
+		return reconcile.Result{}, err
+	}
+	cacheName := instance.Name
+	if instance.Spec.Name != "" {
+		cacheName = instance.Spec.Name
+	}
+	caches, err := cluster.CacheNames(user, pass, podList.Items[0].Name, instance.Namespace, ispnctrl.Protocol(ispnInstance))
+	if err == nil {
+		for i, cn := range caches {
+			if cn == cacheName {
```

Comment:
Can we check that cache is exists via API?
https://infinispan.org/docs/stable/titles/rest/rest.html#rest_v2_cache_exists

Reply:
yeah, good idea. I'll do that

Reply:
done